avail/src/main/kotlin/avail/anvil/shortcuts/KeyboardShortcut.kt

```diff
@@ -0,0 +1,152 @@
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither the name of the copyright holder nor the names of the contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package avail.anvil.shortcuts
+
+import javax.swing.InputMap
+import javax.swing.KeyStroke
+
+/**
+ * The base-level information required to create a keyboard shortcut.
+ *
+ * @author Richard Arriaga
+ */
+interface BaseKeyboardShortcut
+{
+	/**
+	 * The [KeyboardShortcutCategory] this shortcut belongs to.
+	 */
+	val category: KeyboardShortcutCategory
+
+	/**
+	 * The overrideable [Set] of [ModifierKey]s to trigger this shortcut. This
+	 * is the modifier set use to trigger the short cut.
+	 */
+	val modifierKeys: MutableSet<ModifierKey>
+
+	/**
+	 * The [KeyCode] that represents the key when pressed in combination with
+	 * the [modifierKeys] triggers the shortcut. This is the overrideable
+	 * [KeyCode] used to trigger this shortcut
+	 */
+	var keyCode: KeyCode
+
+	/**
+	 * The key that identifies this [BaseKeyboardShortcut].
+	 */
+	val actionMapKey: String
+
+	/**
+	 * The [KeyStroke] used to trigger this [KeyboardShortcut].
+	 */
+	val keyStroke : KeyStroke get() =
+		KeyStroke.getKeyStroke(
+			keyCode.code,
+			modifierKeys.fold(0)
+			{
+					modifier, mk -> modifier.or(mk.modifier)
+			})
+}
+
+/**
+ * An abstract keyboard shortcut used to perform some action in Anvil.
+ *
+ * **NOTE** Keyboard shortcuts must adhere to the following implementation
+ * rules:
+ * 1. All non-abstract subclass implementations must be a child type of a
+ *   sealed class that corresponds to a [KeyboardShortcutCategory].
+ * 2. All non-abstract subclass implementations must be an object.
+ * 3. All abstract subclass implementations must be a sealed class that
+ *   corresponds to a single [KeyboardShortcutCategory].
+ *
+ * @author Richard Arriaga
+ */
+abstract class KeyboardShortcut: BaseKeyboardShortcut
```

Comment:
There were two `open val`'s but `open val description: String = ""` was just a place holder until I could add descriptions to all the shortcuts, which I have, so I switched `description` to `abstract val`s. 

So now we have only one `open val` which is `open val customizable: Boolean = true`. I think I disagree about better readability here. Here' a comparison. In this example I'm pretending we've decided that we don't want the camel case shortcut to be editable. 
In the following block we are leaving the field open, not using the constructor.
`object CamelCaseShortcut: CodePaneShortcut()
{
	override val customizable: Boolean = false
	override val defaultModifierKeys =
		setOf(ModifierKey.META_DOWN_MASK, ModifierKey.CTRL_DOWN_MASK)
	override val defaultKeyCode: KeyCode = KeyCode.VK_C
	override var keyCode = defaultKeyCode
	override val actionMapKey: String = "camel-case"
	override val description: String = "Transform Selected Text to camelCase"
}`

And here we use the abstract constructor:
`object CamelCaseShortcut: CodePaneShortcut(false)
{
	override val defaultModifierKeys =
		setOf(ModifierKey.META_DOWN_MASK, ModifierKey.CTRL_DOWN_MASK)
	override val defaultKeyCode: KeyCode = KeyCode.VK_C
	override var keyCode = defaultKeyCode
	override val actionMapKey: String = "camel-case"
	override val description: String = "Transform Selected Text to camelCase"
}`

I like the first one because you don't have to navigate to the abstract constructor to tell what everything is; it is struct like with all the name fields visible. 

The second example you see a boolean passed to the abstract constructor. It is just a boolean so you have to navigate up the hierarchy to where that constructor introduces the field customizable.

However, if you want me to make this change, I'll do so.

Reply:
But in this example, you could (and probably should if you add the rest of the vals) say
```
object CamelCaseShortcut: CodePaneShortcut(
	customizable = false,
	defaultKey = ...,
	// key = defaultKey,
	actionMapKey = "camel-case",
	description = "Transform Selected Text to camelCase")
```

Reply:
Is there a computational benefit to this? So far as I know this is more personal preference. Right now this is how it looks after my updates to us `Key`:

```kotlin
sealed class CodePaneShortcut constructor(
	override val defaultKey: Key,
	override var key: Key = defaultKey
): KeyboardShortcut()
{
	override val category: KeyboardShortcutCategory
		get() = KeyboardShortcutCategory.CODE_PANE
}

object ExpandTemplateShortcut: CodePaneShortcut(KeyCode.VK_SPACE.with(CTRL))
{
	override val actionMapKey: String = "expand-template"
	override val description: String = "Expand the Selected Template"
}
```

Reply:
It's not critical.  I just prefer not having to say `override val` and state the type and implement the `get()` every time I supply a value for a field.